File: src/screens/SetupScreen.js
import React, { useState } from "react";
import { View, Text, ScrollView, Alert } from "react-native";
import { ScreenHeader, Card, CardRow, SectionHeader, TypeIcon, useC } from "../components/ui";
import { TYPE, SPACE } from "../theme";
import { TYPE_META } from "../data/templates";
import { formatOdo } from "../format";
import { exportJson, exportCsv, exportPdf } from "../data/export";
import {
  builtinConfigs,
  duplicateConfig,
  exportConfig,
  importConfigFile,
} from "../data/config";

export default function SetupScreen({
  vehicles,
  configs,
  onConfigsChange,
  onManage,
  onAddVehicle,
  onCreateFromConfig,
}) {
  const c = useC();
  const [busy, setBusy] = useState(null);
  const [openId, setOpenId] = useState(null);
  const [pdfPick, setPdfPick] = useState(false);

  const saved = configs || [];
  const all = [...builtinConfigs(), ...saved];
  const open = all.find((cfg) => cfg.id === openId);

  const runExport = async (kind, fn) => {
    if (busy) return;
    if (!vehicles.length) {
      Alert.alert("Export", "Add a vehicle first.");
      return;
    }
    setBusy(kind);
    try {
      await fn(vehicles);
    } catch (e) {
      Alert.alert("Export failed", e?.message || String(e));
    } finally {
      setBusy(null);
    }
  };

  const duplicate = (cfg) => {
    const copy = duplicateConfig(cfg);
    onConfigsChange([...saved, copy]);
    setOpenId(copy.id);
  };

  const remove = (cfg) => {
    Alert.alert(`Remove "${cfg.name}"?`, "Vehicles already created from it stay as they are.", [
      { text: "Cancel", style: "cancel" },
      {
        text: "Remove",
        style: "destructive",
        onPress: () => {
          onConfigsChange(saved.filter((x) => x.id !== cfg.id));
          setOpenId(null);
        },
      },
    ]);
  };

  const importFile = async () => {
    try {
      const cfg = await importConfigFile();
      if (!cfg) return;
      onConfigsChange([...saved, cfg]);
      setOpenId(cfg.id);
    } catch (e) {
      Alert.alert("Import failed", e?.message || String(e));
    }
  };

  return (
    <View style={{ flex: 1, backgroundColor: c.bg }}>
      <ScreenHeader title="Setup" onRightPress={onAddVehicle} rightIcon="add" />
      <ScrollView contentContainerStyle={{ paddingHorizontal: SPACE.side, paddingTop: 12, paddingBottom: 40 }}>
        <SectionHeader>Vehicles</SectionHeader>
        {vehicles.length === 0 ? (
          <Text style={[TYPE.small, { color: c.muted, paddingVertical: SPACE.rowY }]}>
            No vehicles yet — tap + to add one.
          </Text>
        ) : (
          <Card>
            {vehicles.map((v, i) => (
              <CardRow
                key={v.id}
                onPress={() => onManage(v.id)}
                chevron
                dot="hidden"
                left={
                  <View style={{ flexDirection: "row", alignItems: "center", gap: 12 }}>
                    <TypeIcon type={v.type} size={32} photo={v.photos?.[0]} />
                    <View>
                      <Text style={[TYPE.title, { color: c.ink }]}>{v.name}</Text>
                      <Text style={[TYPE.small, { color: c.muted, marginTop: 2 }]}>
                        {v.model ? `${v.model}  ·  ` : ""}
                        {TYPE_META[v.type]?.label || v.type}  ·  {formatOdo(v.odometer)} km
                      </Text>
                    </View>
                  </View>
                }
                divider={i < vehicles.length - 1}
              />
            ))}
          </Card>
        )}

        <View style={{ marginTop: SPACE.block }}>
          <SectionHeader>Configs</SectionHeader>
          {open ? (
            <Card>
              <CardRow
                onPress={() => setOpenId(null)}
                dot="hidden"
                left={open.name}
                right="Close"
              />
              <CardRow
                left="Type"
                right={TYPE_META[open.type]?.label || open.type}
                dot="hidden"
              />
              <CardRow
                left="Categories"
                right={String(open.categories.length)}
                dot="hidden"
                divider={open.categories.length > 0}
              />
              {open.categories.map((cat, i) => (
                <CardRow
                  key={`${cat.name}-${i}`}
                  left={cat.name}
                  right={
                    cat.intervalKm
                      ? `${cat.intervalKm} km`
                      : cat.intervalMonths
                        ? `${cat.intervalMonths} mo`
                        : ""
                  }
                  dot="hidden"
                  divider={i < open.categories.length - 1}
                />
              ))}
            </Card>
          ) : null}
          {open ? (
            <Card style={{ marginTop: 12 }}>
              <CardRow
                onPress={() => onCreateFromConfig(open)}
                chevron
                dot="hidden"
                left="Create vehicle"
              />
              <CardRow
                onPress={() => duplicate(open)}
                chevron
                dot="hidden"
                left="Duplicate"
              />
              <CardRow
                onPress={() => exportConfig(open).catch((e) => Alert.alert("Export failed", e?.message || String(e)))}
                chevron
                dot="hidden"
                left="Export"
                divider={!open.builtin}
              />
              {!open.builtin ? (
                <CardRow onPress={() => remove(open)} dot="hidden" left="Delete" divider={false} />
              ) : null}
            </Card>
          ) : (
            <Card>
              {all.map((cfg, i) => (
                <CardRow
                  key={cfg.id}
                  onPress={() => setOpenId(cfg.id)}
                  chevron
                  dot="hidden"
                  left={
                    <View>
                      <Text style={[TYPE.row, { color: c.ink }]}>{cfg.name}</Text>
                      <Text style={[TYPE.small, { color: c.muted, marginTop: 2 }]}>
                        {TYPE_META[cfg.type]?.label || cfg.type}  ·  {cfg.categories.length} categories
                        {cfg.builtin ? "  ·  default" : ""}
                      </Text>
                    </View>
                  }
                  divider={i < all.length - 1}
                />
              ))}
            </Card>
          )}
          {!open ? (
            <Card style={{ marginTop: 12 }}>
              <CardRow onPress={importFile} chevron dot="hidden" left="Import file" divider={false} />
            </Card>
          ) : null}
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <SectionHeader>Export data</SectionHeader>
          <Card>
            <CardRow
              onPress={() => runExport("json", exportJson)}
              chevron
              dot="hidden"
              left="JSON backup"
              right={busy === "json" ? "…" : "Full data"}
            />
            <CardRow
              onPress={() => runExport("csv", exportCsv)}
              chevron
              dot="hidden"
              left="CSV log"
              right={busy === "csv" ? "…" : "Spreadsheet"}
            />
            <CardRow
              onPress={() => setPdfPick((open) => !open)}
              chevron
              dot="hidden"
              left="PDF report"
              right={pdfPick ? "Close" : "Choose"}
              divider={false}
            />
          </Card>
          {pdfPick ? (
            <Card style={{ marginTop: 12 }}>
              <CardRow
                onPress={() => runExport("pdf-all", exportPdf)}
                chevron
                dot="hidden"
                left="All vehicles"
                right={busy === "pdf-all" ? "…" : undefined}
                divider={vehicles.length > 0}
              />
              {vehicles.map((v, i) => (
                <CardRow
                  key={v.id}
                  onPress={() => runExport(`pdf-${v.id}`, () => exportPdf([v]))}
                  chevron
                  dot="hidden"
                  left={v.name}
                  right={busy === `pdf-${v.id}` ? "…" : undefined}
                  divider={i < vehicles.length - 1}
                />
              ))}
            </Card>
          ) : null}
          <Text style={[TYPE.small, { color: c.muted, marginTop: 10, lineHeight: 18 }]}>
            Configs are the checklist setup only. Data export includes logs and odometer.
          </Text>
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <SectionHeader>About</SectionHeader>
          <Card>
            <CardRow left="GreaseTrail" right="1.0" divider={false} dot="hidden" />
          </Card>
        </View>
      </ScrollView>
    </View>
  );
}

File: lib/screens/setup_screen.dart
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

import '../data/config.dart';
import '../data/export.dart';
import '../data/rdw.dart';
import '../format.dart';
import '../models.dart';
import '../theme.dart';
import '../widgets/ui.dart';

class SetupScreen extends StatefulWidget {
  final List<Vehicle> vehicles;
  final List<GtConfig> configs;
  final ValueChanged<List<GtConfig>> onConfigsChange;
  final ValueChanged<String> onManage;
  final VoidCallback onAddVehicle;
  final ValueChanged<GtConfig> onCreateFromConfig;
  final ValueChanged<String> onOpenConfig;
  final Color? accent;
  final ValueChanged<Color?> onAccentChange;
  final bool? rdwSetting;
  final ValueChanged<bool?> onRdwSettingChange;

  const SetupScreen({
    super.key,
    required this.vehicles,
    required this.configs,
    required this.onConfigsChange,
    required this.onManage,
    required this.onAddVehicle,
    required this.onCreateFromConfig,
    required this.onOpenConfig,
    required this.accent,
    required this.onAccentChange,
    required this.rdwSetting,
    required this.onRdwSettingChange,
  });

  @override
  State<SetupScreen> createState() => _SetupScreenState();
}

class _SetupScreenState extends State<SetupScreen> {
  String? _busy;
  bool _pdfPick = false;

  Future<void> _notify(String title, String message) async {
    if (!mounted) return;
    await showDialog<void>(
      context: context,
      builder: (ctx) => AlertDialog(
        title: Text(title),
        content: Text(message),
        actions: [TextButton(onPressed: () => Navigator.pop(ctx), child: const Text('OK'))],
      ),
    );
  }

  Future<void> _runExport(String kind, Future<void> Function(List<Vehicle>, BuildContext) fn) async {
    if (_busy != null) return;
    if (widget.vehicles.isEmpty) {
      await _notify('Export', 'Add a vehicle first.');
      return;
    }
    setState(() => _busy = kind);
    try {
      await fn(widget.vehicles, context);
    } catch (e) {
      await _notify('Export failed', '$e');
    } finally {
      if (mounted) setState(() => _busy = null);
    }
  }

  Future<void> _openKoFi() async {
    final uri = Uri.parse('https://ko-fi.com/lemairetech');
    try {
      final opened = await launchUrl(uri, mode: LaunchMode.externalApplication);
      if (!opened) {
        await _notify('Could not open link', 'Try visiting ko-fi.com/lemairetech in your browser instead.');
      }
    } catch (e) {
      await _notify('Could not open link', '$e');
    }
  }

  Future<void> _importFile() async {
    try {
      final cfg = await importConfigFile();
      if (cfg == null) return;
      widget.onConfigsChange([...widget.configs, cfg]);
      widget.onOpenConfig(cfg.id);
    } catch (e) {
      await _notify('Import failed', '$e');
    }
  }

  @override
  Widget build(BuildContext context) {
    final c = AppColors.of(context);
    final all = [...builtinConfigs(), ...widget.configs];

    return Container(
      color: c.bg,
      child: Column(
        children: [
          ScreenHeader(title: 'Setup', onRightPress: widget.onAddVehicle, rightIcon: Icons.add),
          Expanded(
            child: ListView(
              padding: const EdgeInsets.fromLTRB(AppSpace.side, 12, AppSpace.side, 40),
              children: [
                const SectionHeader('Vehicles'),
                if (widget.vehicles.isEmpty)
                  Padding(
                    padding: const EdgeInsets.symmetric(vertical: AppSpace.rowY),
                    child: Text(
                      'No vehicles yet — tap + to add one.',
                      style: AppTypography.small.copyWith(color: c.muted),
                    ),
                  )
                else
                  AppCard(
                    margin: const EdgeInsets.only(top: 14),
                    children: [
                      for (var i = 0; i < widget.vehicles.length; i++)
                        CardRow(
                          onPress: () => widget.onManage(widget.vehicles[i].id),
                          chevron: true,
                          divider: i < widget.vehicles.length - 1,
                          left: Row(
                            children: [
                              TypeIcon(
                                type: widget.vehicles[i].type,
                                size: 32,
                                photo: widget.vehicles[i].photos.isNotEmpty ? widget.vehicles[i].photos.first : null,
                                color: widget.vehicles[i].color,
                              ),
                              const SizedBox(width: 12),
                              Expanded(
                                child: Column(
                                  crossAxisAlignment: CrossAxisAlignment.start,
                                  children: [
                                    AppText.title(widget.vehicles[i].name, color: c.ink),
                                    Padding(
                                      padding: const EdgeInsets.only(top: 2),
                                      child: AppText.small(
                                        '${widget.vehicles[i].model.isNotEmpty ? '${widget.vehicles[i].model}  ·  ' : ''}'
                                        '${typeMeta[widget.vehicles[i].type]?.label ?? widget.vehicles[i].type}  ·  '
                                        '${formatOdo(widget.vehicles[i].odometer)} km',
                                        color: c.muted,
                                      ),
                                    ),
                                  ],
                                ),
                              ),
                            ],
                          ),
                        ),
                    ],
                  ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('Configs'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    for (var i = 0; i < all.length; i++)
                      CardRow(
                        onPress: () => widget.onOpenConfig(all[i].id),
                        chevron: true,
                        divider: i < all.length - 1,
                        left: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            AppText.row(all[i].name, color: c.ink),
                            Padding(
                              padding: const EdgeInsets.only(top: 2),
                              child: AppText.small(
                                '${typeMeta[all[i].type]?.label ?? all[i].type}  ·  ${all[i].categories.length} categories'
                                '${all[i].builtin ? '  ·  default' : ''}',
                                color: c.muted,
                              ),
                            ),
                          ],
                        ),
                      ),
                  ],
                ),
                AppCard(
                  margin: const EdgeInsets.only(top: 12),
                  children: [CardRow(onPress: _importFile, chevron: true, left: 'Import file', divider: false)],
                ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('Export data'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    CardRow(
                      onPress: () => _runExport('json', exportJson),
                      chevron: true,
                      left: 'JSON backup',
                      right: _busy == 'json' ? '…' : 'Full data',
                    ),
                    CardRow(
                      onPress: () => _runExport('csv', exportCsv),
                      chevron: true,
                      left: 'CSV log',
                      right: _busy == 'csv' ? '…' : 'Spreadsheet',
                    ),
                    CardRow(
                      onPress: () => setState(() => _pdfPick = !_pdfPick),
                      chevron: true,
                      left: 'PDF report',
                      right: _pdfPick ? 'Close' : 'Choose',
                      divider: false,
                    ),
                  ],
                ),
                if (_pdfPick)
                  AppCard(
                    margin: const EdgeInsets.only(top: 12),
                    children: [
                      CardRow(
                        onPress: () => _runExport('pdf-all', exportPdf),
                        chevron: true,
                        left: 'All vehicles',
                        right: _busy == 'pdf-all' ? '…' : null,
                        divider: widget.vehicles.isNotEmpty,
                      ),
                      for (var i = 0; i < widget.vehicles.length; i++)
                        CardRow(
                          onPress: () => _runExport(
                            'pdf-${widget.vehicles[i].id}',
                            (vs, ctx) => exportPdf([widget.vehicles[i]], ctx),
                          ),
                          chevron: true,
                          left: widget.vehicles[i].name,
                          right: _busy == 'pdf-${widget.vehicles[i].id}' ? '…' : null,
                          divider: i < widget.vehicles.length - 1,
                        ),
                    ],
                  ),
                Padding(
                  padding: const EdgeInsets.only(top: 10),
                  child: Text(
                    'Configs are the checklist setup only. Data export includes logs and odometer.',
                    style: AppTypography.small.copyWith(color: c.muted, height: 1.3),
                  ),
                ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('Appearance'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    const Padding(
                      padding: EdgeInsets.fromLTRB(AppSpace.cardPad, 12, AppSpace.cardPad, 0),
                      child: Label('Accent color'),
                    ),
                    ColorSwatchPicker(
                      colors: accentSwatches,
                      value: widget.accent ?? accentSwatches.first,
                      onSelected: (color) => widget.onAccentChange(color == accentSwatches.first ? null : color),
                    ),
                  ],
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 10),
                  child: Text(
                    'Vehicles can override this with their own color — see a vehicle\'s Manage screen.',
                    style: AppTypography.small.copyWith(color: c.muted, height: 1.3),
                  ),
                ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('RDW lookup'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    OptionList<bool?>(
                      options: const [null, true, false],
                      value: widget.rdwSetting,
                      onChanged: widget.onRdwSettingChange,
                      getLabel: (v) => switch (v) {
                        null => 'Automatic',
                        true => 'Always on',
                        false => 'Off',
                      },
                    ),
                  ],
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 10),
                  child: Text(
                    'Shows a license plate lookup when adding a car, motorcycle, or scooter. '
                    "Automatic uses your device's region setting; ${isDutchLocale() ? 'currently on' : 'currently off'}.",
                    style: AppTypography.small.copyWith(color: c.muted, height: 1.3),
                  ),
                ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('About'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    const CardRow(left: 'GreaseTrail', right: '1.2.0'),
                    CardRow(
                      onPress: _openKoFi,
                      chevron: true,
                      left: 'Support the developer',
                      right: 'Ko-fi',
                      divider: false,
                    ),
                  ],
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
